### PR TITLE
[FLINK-20862][table] Support creating DataType from TypeInformation

### PR DESCRIPTION
**Pull request by twalthr**
## What is the purpose of the change

This ensures a smooth mapping between `TypeInformation` and `DataType`. A user can decide whether `TypeInformation` should be treated as an entire RAW type or if instances should be converted to logical types on a best effort basis. In any case, a user will not get an exception during conversion anymore because every `TypeInformation` can be mapped to a non-legacy type in the Blink planner.

## Brief change log

- Add a new TypeInformation to DataType converter that is backed by `DataTypeFactory`
- Add API methods for creating a `DataType` either via `DataTypes.of(TypeInformation)` or `DataTypes.RAW(TypeInformation)`

## Verifying this change

This change added tests and can be verified as follows:

- `DataTypesTest`
- `TypeInfoDataTypeConverterTest`
- `TypeInfoDataTypeConverterScalaTest`

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): yes
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: yes
  - The serializers: no
  - The runtime per-record code paths (performance sensitive): no
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn/Mesos, ZooKeeper: no
  - The S3 file system connector: no

## Documentation

  - Does this pull request introduce a new feature? yes
  - If yes, how is the feature documented? JavaDocs

### CODE REVIEW

File: flink-table/flink-table-common/src/test/java/org/apache/flink/table/types/TypeInfoDataTypeConverterTest.java [hunk 1]
```diff
@@ -0,0 +1,262 @@
...
+                                                                                DataTypes.STRING()
+                                                                                        .nullable()),
+                                                                        DataTypes.FIELD(
+                                                                                "f1",
+                                                                                DataTypes.BOOLEAN()
+                                                                                        .notNull()))
+                                                                .notNull()),
+                                                DataTypes.FIELD("f2", DataTypes.BIGINT().notNull()))
+                                        .notNull()),
+                TestSpec.forType(Types.POJO(PojoWithFieldOrder.class))
```
Comment (dawidwys) [inline flink-table/flink-table-common/src/test/java/org/apache/flink/table/types/TypeInfoDataTypeConverterTest.java:92]: I think something is wrong with POJO types mapping. Shouldn't the primitive fields be mapped to `NOT NULL` types?
Also from our offline chat it seems bridging classes are also not mapped correctly.

**Review by dawidwys — CHANGES_REQUESTED** [review]

**Comment by twalthr** [comment]
Thanks for the feedback @dawidwys. I have updated the PR.

File: flink-table/flink-table-common/src/main/java/org/apache/flink/table/types/utils/TypeInfoDataTypeConverter.java [hunk 2]
```diff
@@ -260,35 +261,70 @@ private static DataType convertToStructuredType(
...
+
+            // based on type information all fields are boxed classes,
+            // therefore we need to check the reflective field for more details
+            fieldDataTypes.replaceAll(
+                    (name, dataType) -> {
+                        final Class<?> fieldClass =
+                                pojoFields.stream()
+                                        .filter(f -> f.getName().equals(name))
+                                        .findFirst()
+                                        .orElseThrow(IllegalArgumentException::new)
```
Comment (dawidwys) [inline flink-table/flink-table-common/src/main/java/org/apache/flink/table/types/utils/TypeInfoDataTypeConverter.java:286]: nit: I'd rather use `IllegalStateException` as it is more likely an internal Flink error that the `TypeInformation` does not reflect its class.

File: flink-table/flink-table-common/src/main/java/org/apache/flink/table/types/utils/TypeInfoDataTypeConverter.java [hunk 3]
```diff
@@ -260,35 +261,70 @@ private static DataType convertToStructuredType(
...
+                                        .getType();
+                        if (fieldClass.isPrimitive()) {
+                            return dataType.bridgedTo(fieldClass);
```
Comment (dawidwys) [inline flink-table/flink-table-common/src/main/java/org/apache/flink/table/types/utils/TypeInfoDataTypeConverter.java:289]: nit: How about we call `notNull` here, just for clarity.

**Review by dawidwys — APPROVED** [review]
Looks good now!

**Pull request closed without merging**